dmidecode.py

```diff
@@ -0,0 +1,152 @@
+    """
+    Parse the whole dmidecode output.
+    Returns a list of tuples of (type int, value dict).
+    """
+    info = []
+    lines = iter(content.strip().splitlines())
+    while True:
+        try:
+            line = next(lines)
+        except StopIteration:
+            break
+
+        if line.startswith('Handle 0x'):
+            typ = int(line.split(',', 2)[1].strip()[len('DMI type'):])
+            if typ in TYPE:
+                info.append((TYPE[typ], _parse_handle_section(lines)))
+    return info
+
+
+def _parse_handle_section(lines):
+    """
+    Parse a section of dmidecode output
+
+    * 1st line contains address, type and size
+    * 2nd line is title
+    * line started with one tab is one option and its value
+    * line started with two tabs is a member of list
+    """
+    data = {
+        '_title': next(lines).rstrip(),
+        }
+
+    for line in lines:
+        line = line.rstrip()
+        if line.startswith('\t\t'):
+            if isinstance(data[k], list):
+                data[k].append(line.lstrip())
+        elif line.startswith('\t'):
+            k, v = [i.strip() for i in line.lstrip().split(':', 1)]
+            if v:
+                data[k] = v
+            else:
+                data[k] = []
+        else:
+            break
+
+    return data
+
+
+def profile():
+#    if os.isatty(sys.stdin.fileno()):
+#        content = _get_output()
+#    else:
+#        content = sys.stdin.read()
+    content = _get_output()
+
+    info = parse_dmi(content)
+    _show(info)
+
+
+def _get_output():
+    import subprocess
+    try:
+        output = subprocess.check_output('/usr/sbin/dmidecode', shell=False)
+
+    except Exception as e:
+        print(e, file=sys.stderr)
+        if str(e).find("command not found") == -1:
+            print("please install dmidecode", file=sys.stderr)
+            print("e.g. sudo apt install dmidecode",file=sys.stderr)
+
+        sys.exit(1)
+    return output.decode()
+
+#
+# Builds a label dict from o, with all the 'expected' labels reduced to lowercase-letter-only labels
+# and filtering out those whose values are listed in 'rejected'
+#
+def _get_labels(o, expected, rejected):
+    r = {}
+    for l in expected:
+        if l in o.keys():
+            if o[l] not in rejected:
+                r[''.join(filter(str.isalpha, l.lower()))] = o[l]
+    return r
+
+def _show(info):
+    def _get(i):
+        return [v for j, v in info if j == i]
+
+    system = _get('system')[0]
+    l = _get_labels(system, ['Manufacturer','Product Name', 'Serial Number'], ['Not Specified'])
+    print ('node_dmi_hardware_info{{{}}} 1'.format( ','.join('{}="{}"'.format(k,v) for k,v in l.items()) ))
+
+    idx = 0
+    for cpu in _get('processor'):
+        l = _get_labels(cpu, ['Manufacturer','Family', 'Max Speed', 'Core Count'], ['Not Specified'])
+        l['index'] = idx
+        print ('node_dmi_processor{{{}}} 1'.format( ','.join('{}="{}"'.format(k,v) for k,v in l.items()) ))
```

Comment:
Since these are all info metrics, they should have consistent `_info` in the names.